### ISSUE TITLE
Example of LSTM code and possible regression with batchfirst

### ISSUE THREAD

**Issue by ceberly** [soumith/cudnn.torch#209]
Hey I hope this isn't too off topic. I've been playing around with the LSTM code and I can't get it to work. I've plugged into the ```torch-rnn``` project like this:

```
...
  local V, D, H = self.vocab_size, self.wordvec_dim, self.rnn_size

  self.net = nn.Sequential()
  self.rnns = {}
  self.bn_view_in = {}
  self.bn_view_out = {}

  self.net:add(nn.LookupTable(V, D))

  local rnn = cudnn.LSTM(D, H, self.num_layers, true)
..
```

Is there an example of the LSTM code in action somewhere? Nvidia is woefully short on documentation. It seems like the test step converges fine, but eval losses are all over the place. Also since the most recent pull of #207 I can't get testing to work at all.

Here's an example using the tiny-shakespeare code:

```
chris@alpha:~/src/torch-rnn$ time th train.lua -input_h5 ..data/tiny-shakespeare.h5 -input_json ../data/tiny-shakespeare.json  -max_epochs 5 -print_every 100 -model_type culstm
Running with CUDA on GPU 0
Epoch 1.28 / 5, i = 100 / 1780, loss = 3.438158
Epoch 1.56 / 5, i = 200 / 1780, loss = 3.580820
Epoch 1.84 / 5, i = 300 / 1780, loss = 3.411032
Epoch 2.12 / 5, i = 400 / 1780, loss = 3.456796
Epoch 2.40 / 5, i = 500 / 1780, loss = 3.494461
```

Previously the loss went down quite dramatically, and regressing to the original lstm code shows a similar effect.

Anyway i'm excited about this integration, any help is appreciated!

**Comment by ngimel** [comment]
There are some issues with the batchFirst. Until they are sorted out, you can try the following workaround:
```
rnn = cudnn.LSTM(D, H, self.num_layers, false)
self.net:add(nn.Transpose({1, 2}))
self.net:add(nn.Contiguous())
self.net:add(rnn)
self.net:add(nn.Transpose({1, 2}))
self.net:add(nn.Contiguous())
```

**Comment by ceberly** [comment]
thank you, now that I've located the user guide this should be a lot easier.

**Comment by SeanNaren** [comment]
I think an example of how the RNN code could be plugged into a simple net training on artificial data wouldn't be too difficult. Any ideas on how we could approach this, and where it could be kept?

**Comment by ceberly** [comment]
I'm working on something that i should have done today. I would just have to remove the stuff that's specific to my company and then I can follow up here. Its basically just nuts and bolts of torch-rnn but using cudnn. I'll report back!

**Comment by ngimel** [comment]
A working example of using cudnn LSTM with torch-rnn is here
https://github.com/ngimel/torch-rnn

**Comment by ceberly** [comment]
@ngimel I know this issue is closed but I just wanted to add that this code only works with cudnn > 5.1 which is of course fine but I'm leaving it here for posterity. cudnn 5 training doesn't seem to converge.

Thank you! Here are the results on tiny-shakespeare

With cudnn5:
```
chris@alpha:~/src/torch-rnn$ th train.lua -max_epochs 2 -print_every 100
Running with CUDA on GPU 0
Epoch 1.28 / 2, i = 100 / 712, loss = 2.420157
Epoch 1.56 / 2, i = 200 / 712, loss = 2.146538
Epoch 1.84 / 2, i = 300 / 712, loss = 1.986639
Epoch 2.12 / 2, i = 400 / 712, loss = 1.863283
Epoch 2.40 / 2, i = 500 / 712, loss = 1.836988
Epoch 2.69 / 2, i = 600 / 712, loss = 1.766764
Epoch 2.97 / 2, i = 700 / 712, loss = 1.771655
val_loss =      1.8160738186403
chris@alpha:~/src/torch-rnn$ th train.lua -max_epochs 2 -print_every 100 -cudnn 1
Running with CUDA on GPU 0
Epoch 1.28 / 2, i = 100 / 712, loss = 3.439970
Epoch 1.56 / 2, i = 200 / 712, loss = 3.569918
Epoch 1.84 / 2, i = 300 / 712, loss = 3.384764
Epoch 2.12 / 2, i = 400 / 712, loss = 3.463539
Epoch 2.40 / 2, i = 500 / 712, loss = 3.501738
Epoch 2.69 / 2, i = 600 / 712, loss = 3.402064
Epoch 2.97 / 2, i = 700 / 712, loss = 3.467268
val_loss =      3.2156173749404
```

with cudnn 5.1
```
chris@alpha:~/src/torch-rnn$ time th train.lua -max_epochs 2 -print_every 100 -cudnn 0
Running with CUDA on GPU 0
Epoch 1.28 / 2, i = 100 / 712, loss = 2.424212
Epoch 1.56 / 2, i = 200 / 712, loss = 2.157564
Epoch 1.84 / 2, i = 300 / 712, loss = 2.002452
Epoch 2.12 / 2, i = 400 / 712, loss = 1.873721
Epoch 2.40 / 2, i = 500 / 712, loss = 1.855224
Epoch 2.69 / 2, i = 600 / 712, loss = 1.770854
Epoch 2.97 / 2, i = 700 / 712, loss = 1.772905
val_loss =      1.8264290636236

real    0m52.774s
user    0m47.204s
sys     0m5.624s
chris@alpha:~/src/torch-rnn$ time th train.lua -max_epochs 2 -print_every 100 -cudnn 1
Running with CUDA on GPU 0
Epoch 1.28 / 2, i = 100 / 712, loss = 2.396797
Epoch 1.56 / 2, i = 200 / 712, loss = 2.119369
Epoch 1.84 / 2, i = 300 / 712, loss = 1.977171
Epoch 2.12 / 2, i = 400 / 712, loss = 1.830874
Epoch 2.40 / 2, i = 500 / 712, loss = 1.816489
Epoch 2.69 / 2, i = 600 / 712, loss = 1.748117
Epoch 2.97 / 2, i = 700 / 712, loss = 1.748035
val_loss =      1.8025269535455
```

Hope this helps people

**Comment by ngimel** [comment]
Stange... Here's my output with cudnn v5 (I've added print('cudnn version', cudnn.version) in the beginning):
```
ngimel@ngimel-LT ~/ALL/torch/torch-rnn.ngimel $ th train.lua -input_h5 tiny-shakespeare.h5 -input_json tiny-shakespeare.json -cudnn 1 -print_every 100 -max_epochs 2
cudnn version	5005	
Running with CUDA on GPU 0	
Epoch 1.28 / 2, i = 100 / 712, loss = 2.438138	
Epoch 1.56 / 2, i = 200 / 712, loss = 2.152842	
Epoch 1.84 / 2, i = 300 / 712, loss = 1.990458	
Epoch 2.12 / 2, i = 400 / 712, loss = 1.846732	
Epoch 2.40 / 2, i = 500 / 712, loss = 1.824926	
Epoch 2.69 / 2, i = 600 / 712, loss = 1.752210	
Epoch 2.97 / 2, i = 700 / 712, loss = 1.764060	
val_loss = 	1.8133310702714	
```

**Comment by ceberly** [comment]
Weird. Well anyway thank you, it's much faster now and I'm getting reasonable results!
